File: packages/sdk-go/plugins/paymaster/verifying_paymaster.go
package paymaster

import (
	"context"
	"crypto/ecdsa"
	"fmt"
	"math/big"
	"time"

	"github.com/ethereum/go-ethereum/crypto"

	"github.com/stablenet/sdk-go/core"
	corepaymaster "github.com/stablenet/sdk-go/core/paymaster"
	"github.com/stablenet/sdk-go/types"
)

// ============================================================================
// Verifying Paymaster Configuration
// ============================================================================

// VerifyingPaymasterConfig configures the verifying paymaster.
type VerifyingPaymasterConfig struct {
	// PaymasterAddress is the paymaster contract address.
	PaymasterAddress types.Address
	// EntryPointAddress is the EntryPoint contract address (default: ERC-4337 v0.7).
	EntryPointAddress *types.Address
	// PrivateKey is the signer's private key for signing paymaster data.
	PrivateKey types.Hex
	// ChainID is the chain ID.
	ChainID types.ChainID
	// ValiditySeconds is the validity window in seconds (default: 3600 = 1 hour).
	ValiditySeconds int
}

// Default gas limits for verifying paymaster operations.
const (
	DefaultVerifyingPaymasterVerificationGas = 100_000
	DefaultVerifyingPaymasterPostOpGas       = 50_000
	DefaultValiditySeconds                   = 3600
)

// ============================================================================
// Verifying Paymaster
// ============================================================================

// VerifyingPaymaster uses off-chain signature to approve gas sponsorship.
type VerifyingPaymaster struct {
	config          VerifyingPaymasterConfig
	privateKey      *ecdsa.PrivateKey
	entryPoint      types.Address
	domainSeparator types.Hash
}

// NewVerifyingPaymaster creates a new verifying paymaster.
func NewVerifyingPaymaster(cfg VerifyingPaymasterConfig) (*VerifyingPaymaster, error) {
	if len(cfg.PrivateKey) == 0 {
		return nil, fmt.Errorf("private key is required")
	}

	// Parse private key
	privateKey, err := crypto.ToECDSA(cfg.PrivateKey.Bytes())
	if err != nil {
		return nil, fmt.Errorf("invalid private key: %w", err)
	}

	if cfg.ValiditySeconds == 0 {
		cfg.ValiditySeconds = DefaultValiditySeconds
	}

	entryPoint := core.EntryPointV07Address
	if cfg.EntryPointAddress != nil {
		entryPoint = *cfg.EntryPointAddress
	}

	// Pre-compute domain separator (immutable for given chain/entrypoint/paymaster)
	chainID := new(big.Int).SetUint64(uint64(cfg.ChainID))
	domainSeparator := corepaymaster.ComputeDomainSeparator(chainID, entryPoint, cfg.PaymasterAddress)

	return &VerifyingPaymaster{
		config:          cfg,
		privateKey:      privateKey,
		entryPoint:      entryPoint,
		domainSeparator: domainSeparator,
	}, nil
}

// GetPaymasterStubData returns stub data for gas estimation.
func (v *VerifyingPaymaster) GetPaymasterStubData(ctx context.Context, userOp *types.PartialUserOperation, cfg PaymentConfig) (*StubData, error) {
	// Build envelope with placeholder timestamps and empty payload
	validUntil := uint64(time.Now().Unix()) + uint64(v.config.ValiditySeconds)
	validAfter := uint64(0)

	envelope, err := corepaymaster.EncodePaymasterData(&corepaymaster.Envelope{
		PaymasterType: corepaymaster.PaymasterTypeVerifying,
		Flags:         0,
		ValidUntil:    validUntil,
		ValidAfter:    validAfter,
		Nonce:         0,
		Payload:       nil,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to encode stub envelope: %w", err)
	}

	// Append a 65-byte stub signature
	stubSig := make([]byte, 65)
	paymasterData := corepaymaster.ConcatWithSignature(envelope, stubSig)

	return &StubData{
		Paymaster:                     v.config.PaymasterAddress,
		PaymasterData:                 types.Hex(paymasterData),
		PaymasterVerificationGasLimit: big.NewInt(DefaultVerifyingPaymasterVerificationGas),
		PaymasterPostOpGasLimit:       big.NewInt(DefaultVerifyingPaymasterPostOpGas),
	}, nil
}

// GetPaymasterData returns the final paymaster data with signature.
func (v *VerifyingPaymaster) GetPaymasterData(ctx context.Context, userOp *types.PartialUserOperation, cfg PaymentConfig) (*PaymasterData, error) {
	// Set validity window
	validUntil := uint64(time.Now().Unix()) + uint64(v.config.ValiditySeconds)
	validAfter := uint64(0)

	// Build v2 envelope (without signature)
	envelope, err := corepaymaster.EncodePaymasterData(&corepaymaster.Envelope{
		PaymasterType: corepaymaster.PaymasterTypeVerifying,
		Flags:         0,
		ValidUntil:    validUntil,
		ValidAfter:    validAfter,
		Nonce:         0,
		Payload:       nil,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to encode envelope: %w", err)
	}

	// Compute user op core hash
	nonce := big.NewInt(0)
	if userOp.Nonce != nil {
		nonce = userOp.Nonce
	}
	preVerificationGas := big.NewInt(0)
	if userOp.PreVerificationGas != nil {
		preVerificationGas = userOp.PreVerificationGas
	}

	// Build initCode from factory + factoryData
	var initCode []byte
	if userOp.Factory != nil {
		initCode = append(initCode, userOp.Factory.Bytes()...)
		initCode = append(initCode, userOp.FactoryData.Bytes()...)
	}

	accountGasLimits := corepaymaster.PackGasLimits(userOp.VerificationGasLimit, userOp.CallGasLimit)
	gasFees := corepaymaster.PackGasLimits(userOp.MaxPriorityFeePerGas, userOp.MaxFeePerGas)

	userOpCoreHash := corepaymaster.ComputeUserOpCoreHash(&corepaymaster.UserOpCoreFields{
		Sender:             userOp.Sender,
		Nonce:              nonce,
		InitCode:           initCode,
		CallData:           userOp.CallData.Bytes(),
		AccountGasLimits:   accountGasLimits,
		PreVerificationGas: preVerificationGas,
		GasFees:            gasFees,
	})

	// Compute the hash that needs to be signed
	paymasterHash := corepaymaster.ComputePaymasterHash(v.domainSeparator, userOpCoreHash, envelope)

	// Sign the hash
	signature, err := crypto.Sign(paymasterHash.Bytes(), v.privateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to sign paymaster hash: %w", err)
	}

	// Concatenate envelope + signature
	paymasterData := corepaymaster.ConcatWithSignature(envelope, signature)

	return &PaymasterData{
		Paymaster:                     v.config.PaymasterAddress,
		PaymasterData:                 types.Hex(paymasterData),
		PaymasterVerificationGasLimit: big.NewInt(DefaultVerifyingPaymasterVerificationGas),
		PaymasterPostOpGasLimit:       big.NewInt(DefaultVerifyingPaymasterPostOpGas),
	}, nil
}

// GetSupportedTokens returns empty list (verifying paymaster doesn't support tokens).
func (v *VerifyingPaymaster) GetSupportedTokens(ctx context.Context) ([]TokenInfo, error) {
	return nil, nil
}

// EstimateTokenPayment returns error (verifying paymaster doesn't support tokens).
func (v *VerifyingPaymaster) EstimateTokenPayment(ctx context.Context, userOp *types.PartialUserOperation, token types.Address) (*TokenEstimate, error) {
	return nil, fmt.Errorf("verifying paymaster does not support token payments")
}

// IsSponsorshipAvailable always returns true for verifying paymaster.
func (v *VerifyingPaymaster) IsSponsorshipAvailable(ctx context.Context, sender types.Address) (bool, error) {
	return true, nil
}

// GetSponsorshipPolicy returns a basic policy.
func (v *VerifyingPaymaster) GetSponsorshipPolicy(ctx context.Context, sender types.Address) (*SponsorshipPolicy, error) {
	return &SponsorshipPolicy{
		IsAvailable: true,
	}, nil
}

// ============================================================================
// Helper Functions
// ============================================================================

// stubSignature returns a 65-byte stub signature for gas estimation.
func stubSignature() types.Hex {
	return make(types.Hex, 65)
}

// DecodeVerifyingPaymasterData decodes verifying paymaster data using the v2 envelope format.
// Returns the decoded envelope and the trailing signature.
func DecodeVerifyingPaymasterData(data types.Hex) (*corepaymaster.Envelope, types.Hex, error) {
	envelope, sig, err := corepaymaster.SplitEnvelopeAndSignature(data.Bytes(), 65)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to split envelope and signature: %w", err)
	}

	env, err := corepaymaster.DecodePaymasterData(envelope)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to decode envelope: %w", err)
	}

	return env, types.Hex(sig), nil
}

File: packages/sdk-ts/plugins/paymaster/src/verifyingPaymaster.ts
import type {
  PaymasterClient,
  PaymasterData,
  PaymasterStubData,
  UserOperation,
} from '@stablenet/sdk-types'
import type { Address, Hex } from 'viem'
import { concat, pad, toHex } from 'viem'
import {
  encodePaymasterData,
  encodePaymasterDataWithSignature,
  PaymasterType,
  computePaymasterDomainSeparator,
  computeUserOpCoreHash,
  computePaymasterHash,
  encodeVerifyingPayload,
} from '@stablenet/core'
import type { VerifyingPaymasterConfig } from './types'
import { DEFAULT_VALIDITY_SECONDS } from './types'

// Default gas limits for paymaster operations
const DEFAULT_PAYMASTER_VERIFICATION_GAS_LIMIT = 100_000n
const DEFAULT_PAYMASTER_POST_OP_GAS_LIMIT = 50_000n

// Stub signature for gas estimation (65 bytes of zeros)
const STUB_SIGNATURE: Hex = `0x${'00'.repeat(65)}`

/**
 * Pack a UserOperation's fields into the format needed by computeUserOpCoreHash.
 * Converts unpacked v0.7 UserOperation fields into packed Hex fields.
 */
function packUserOpForHash(userOp: UserOperation): {
  sender: Address
  nonce: bigint
  initCode: Hex
  callData: Hex
  accountGasLimits: Hex
  preVerificationGas: bigint
  gasFees: Hex
} {
  const initCode =
    userOp.factory && userOp.factoryData
      ? concat([userOp.factory, userOp.factoryData])
      : ('0x' as Hex)

  const accountGasLimits = concat([
    pad(toHex(userOp.verificationGasLimit), { size: 16 }),
    pad(toHex(userOp.callGasLimit), { size: 16 }),
  ]) as Hex

  const gasFees = concat([
    pad(toHex(userOp.maxPriorityFeePerGas), { size: 16 }),
    pad(toHex(userOp.maxFeePerGas), { size: 16 }),
  ]) as Hex

  return {
    sender: userOp.sender,
    nonce: userOp.nonce,
    initCode,
    callData: userOp.callData,
    accountGasLimits,
    preVerificationGas: userOp.preVerificationGas,
    gasFees,
  }
}

/**
 * Create a Verifying Paymaster client
 *
 * The VerifyingPaymaster uses an off-chain signature to approve gas sponsorship.
 * The signer signs a hash of the user operation data to determine which operations to sponsor.
 *
 * Data is encoded in envelope format (version byte 0x01) with type-specific payloads.
 *
 * @example
 * ```ts
 * import { createVerifyingPaymaster } from '@stablenet/plugin-paymaster'
 * import { privateKeyToAccount } from 'viem/accounts'
 *
 * const signer = privateKeyToAccount('0x...')
 * const paymaster = createVerifyingPaymaster({
 *   paymasterAddress: '0x...',
 *   signer,
 *   chainId: 1n,
 * })
 * ```
 */
export function createVerifyingPaymaster(config: VerifyingPaymasterConfig): PaymasterClient {
  const { paymasterAddress, signer, chainId, validitySeconds = DEFAULT_VALIDITY_SECONDS } = config

  /**
   * Get stub data for gas estimation
   * Uses placeholder signature since actual signature isn't needed for estimation
   */
  const getPaymasterStubData = async (
    _userOperation: UserOperation,
    entryPoint: Address,
    _chainId: bigint
  ): Promise<PaymasterStubData> => {
    const validUntil = BigInt(Math.floor(Date.now() / 1000) + validitySeconds)
    const validAfter = 0n

    // Build an empty verifying payload (no policy for stub)
    const payload = encodeVerifyingPayload({
      policyId: '0x0000000000000000000000000000000000000000000000000000000000000000' as Hex,
      sponsor: '0x0000000000000000000000000000000000000000' as Address,
      maxCost: 0n,
      verifierExtra: '0x',
    })

    // Encode the envelope
    const envelope = encodePaymasterData({
      paymasterType: PaymasterType.VERIFYING,
      flags: 0,
      validUntil,
      validAfter,
      nonce: 0n,
      payload,
    })

    // Concatenate envelope with stub signature
    const paymasterData = encodePaymasterDataWithSignature(envelope, STUB_SIGNATURE)

    return {
      paymaster: paymasterAddress,
      paymasterData,
      paymasterVerificationGasLimit: DEFAULT_PAYMASTER_VERIFICATION_GAS_LIMIT,
      paymasterPostOpGasLimit: DEFAULT_PAYMASTER_POST_OP_GAS_LIMIT,
    }
  }

  /**
   * Get paymaster data with actual signature
   */
  const getPaymasterData = async (
    userOperation: UserOperation,
    entryPoint: Address,
    _chainId: bigint
  ): Promise<PaymasterData> => {
    // Set validity window (configurable, default: 1 hour)
    const validUntil = BigInt(Math.floor(Date.now() / 1000) + validitySeconds)
    const validAfter = 0n

    // Build verifying payload
    const payload = encodeVerifyingPayload({
      policyId: '0x0000000000000000000000000000000000000000000000000000000000000000' as Hex,
      sponsor: '0x0000000000000000000000000000000000000000' as Address,
      maxCost: 0n,
      verifierExtra: '0x',
    })

    // Encode the envelope (without signature)
    const envelope = encodePaymasterData({
      paymasterType: PaymasterType.VERIFYING,
      flags: 0,
      validUntil,
      validAfter,
      nonce: 0n,
      payload,
    })

    // Compute the hash using the 3-step core approach
    const domainSeparator = computePaymasterDomainSeparator(
      chainId,
      entryPoint,
      paymasterAddress
    )
    const packedOp = packUserOpForHash(userOperation)
    const userOpCoreHash = computeUserOpCoreHash(packedOp)
    const hash = computePaymasterHash(domainSeparator, userOpCoreHash, envelope)

    // Sign the hash
    const signature = await signer.signMessage({
      message: { raw: hash },
    })

    // Encode paymaster data with real signature
    const paymasterData = encodePaymasterDataWithSignature(envelope, signature)

    return {
      paymaster: paymasterAddress,
      paymasterData,
    }
  }

  return {
    getPaymasterStubData,
    getPaymasterData,
  }
}

/**
 * Create a verifying paymaster from a private key
 */
export async function createVerifyingPaymasterFromPrivateKey(config: {
  paymasterAddress: Address
  privateKey: Hex
  chainId: bigint
  entryPoint?: Address
}): Promise<PaymasterClient> {
  const { privateKeyToAccount } = await import('viem/accounts')
  const signer = privateKeyToAccount(config.privateKey)

  return createVerifyingPaymaster({
    paymasterAddress: config.paymasterAddress,
    signer,
    chainId: config.chainId,
    entryPoint: config.entryPoint,
  })
}
